chained/water_magnet_downloader.go

```diff
@@ -0,0 +1,52 @@
+package chained
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/anacrolix/torrent"
+)
+
+type magnetDownloader struct {
+	magnetURL string
+}
+
+// NewMagnetDownloader creates a new WASMDownloader instance.
+func NewMagnetDownloader(magnetURL string) WASMDownloader {
+	return &magnetDownloader{
+		magnetURL: magnetURL,
+	}
+}
+
+// DownloadWASM downloads the WASM file from the given URL.
+func (d *magnetDownloader) DownloadWASM(ctx context.Context, w io.Writer) error {
+	cfg := torrent.NewDefaultClientConfig()
+	path, err := os.MkdirTemp("/tmp", "torrent")
+	if err != nil {
+		return fmt.Errorf("failed to create temp dir: %w", err)
+	}
+	cfg.DataDir = path
+	client, err := torrent.NewClient(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create torrent client: %w", err)
+	}
+	defer client.Close()
+
+	t, err := client.AddMagnet(d.magnetURL)
+	if err != nil {
+		return fmt.Errorf("failed to add magnet: %w", err)
+	}
+	<-t.GotInfo()
+	t.DownloadAll()
+	client.WaitAll()
```

Comment:
This looks like it will block potentially indefinitely and therefore not honor the provided context.  However, it looks like you can configure an `HTTPDialContext` on the client.  I'm unclear on whether that will apply to this call.  There is also a [`TrackerDialContext`](https://pkg.go.dev/github.com/anacrolix/torrent#ClientTrackerConfig).

Reply:
If those context fields don't apply, there is also [CancelPieces](https://pkg.go.dev/github.com/anacrolix/torrent#Torrent.CancelPieces).  You could call `t.DownloadAll()` and `client.WaitAll` in a separate goroutine, then in the main routine have something like:
```
select {
case <-downloadComplete:
case <-ctx.Done():
    t.CancelPieces(0, t.NumPieces())
}
```

Reply:
> This looks like it will block potentially indefinitely and therefore not honor the provided context. However, it looks like you can configure an `HTTPDialContext` on the client. I'm unclear on whether that will apply to this call. There is also a [`TrackerDialContext`](https://pkg.go.dev/github.com/anacrolix/torrent#ClientTrackerConfig).

Wasn't aware about that, it should make it easier to test too, thanks!

Reply:
You might want to use NewReader on the Torrent, and just stream it straight out to the writer `w` here. There is no provided event to wait for completed downloads in anacrolix/torrent, although it's a commonly requested feature. I prefer to point people toward the streaming reader support.